### ISSUE TITLE
Allow for sticky keyword routing.

### ISSUE THREAD

**Issue by smn** [praekeltfoundation/vumi-go#956]
1. It should be optional and default to `False`
2. When turned on it should make routing sticky until:
   1. The application sends a `session end`
   2. The transport sends a `session close`
   3. A UI configurable timeout is reached.

**Comment by smn** [comment]
@hodgestar @justinvdm @badlard thoughts?

**Comment by hodgestar** [comment]
The first message we receive for USSD is usually a `session_new` with
`null` content. I don't know how the keyword router could route that or
what response it would send if it didn't. Maybe session-based transports
are better handled by the `ApplicationMultiplexer`?

For non-session transports I think this makes sense.

**Comment by smn** [comment]
This would be for non-session transports but on a message level we don't have means yet to detect which one we're dealing with. Isn't this a problem already anyway for people hooking up session based transports to a keyword router or do we already have ways of preventing that?

**Comment by hodgestar** [comment]
Yep, it's already a problem. Happy for us to just ignore the issue of
sessions when we add this to the keyword router.

**Comment by badlard** [comment]
Yup it's a problem and the reason why I am still telling business/strategy folks to not (yet) use the keyword router in the UI.

**Comment by smn** [comment]
So what's the way forward on this to make it workable? Not use it with session based transports at all or make it sticky by default for all things session-y? A piece of work I'm looking at now needs this.

**Comment by jerith** [comment]
The keyword router was designed specifically for SMS and was never intended to be used with session-based channels. I think we'd be better off figuring out what functionality we need and building a new router to do that.

**Comment by smn** [comment]
The request this is coming from is for Mxit:
1. A bulk message is sent out and people are asked to reply with a certain keyword
2. The keyword should go back to the keyword router which routes it to a specific jsbox app
3. Subsequent replies to that jsbox app shouldn't need to be prefixed with the keyword, it should be "sticky" until the app sends a session close message.

If the keyword router isn't the place then what is, what would we call it and how would it work differently?

**Comment by jerith** [comment]
This sounds like something a lot more like the app mux router. That currently does menu-based app selection, but we could either extend it to support keywords or build a new keyword-based version.

**Comment by hodgestar** [comment]
For my part, I'm happy with this being added to the keyword router (sorry
if that wasn't clear from my existing comment).

Having the app be able to trigger an end to the stickiness with a
session_event `close` sounds sensible. I don't think the session_close
from the transport would ever be used, but I don't mind having it in since
it seems sensible.

**Comment by badlard** [comment]
In terms of the keyword router - it must be sticky for (at minimum)
non-session based channels. From a product perspective, my preference would
be to improve the current keyword router

On 20 May 2014 16:24, Jeremy Thurgood notifications@github.com wrote:

> This sounds like something a lot more like the app mux router. That
> currently does menu-based app selection, but we could either extend it to
> support keywords or build a new keyword-based version.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/praekelt/vumi-go/issues/956#issuecomment-43632267
> .

## 

_BEN ADLARD_ | PRODUCT MANAGER

_PRAEKELTFOUNDATION_.ORG
_M_:  +27823182071  |  _S_: badlard
_T_: vumigo.tumblr.com  |  *T: *@vumiapp https://twitter.com/vumiapp
